### ISSUE TITLE
Module parse failed: Unexpected token when transpiling with webpack due to inclusion of TypeScript optional chaining operator (`submitter?.formMethod`)

### ISSUE THREAD

**Issue by davidrunger** [hotwired/turbo-rails#336]
I am attempting to bump @hotwired/turbo-rails from 7.1.1 to 7.1.2, but I am getting this error when running `bin/webpack` in my Rails app:

```
ERROR in ./node_modules/@hotwired/turbo-rails/app/javascript/turbo/form_submissions.js 9:31
Module parse failed: Unexpected token (9:31)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|   }
| }) {
>   const formMethod = submitter?.formMethod;
|
|   if (formMethod && fetchRequest.body.has("_method")) {
 @ ./node_modules/@hotwired/turbo-rails/app/javascript/turbo/index.js 2:0-66 7:39-67
 @ ./app/javascript/packs/quizzes.js
```

It seems that the error is being caused by the inclusion of the TypeScript optional chaining operator (`?.`) in that file (`app/javascript/turbo/form_submissions.js`) in `submitter?.formMethod`.

I think that this issue was introduced in https://github.com/hotwired/turbo-rails/pull/239 (Support `[formmethod]` overrides to `_method`) here: https://github.com/hotwired/turbo-rails/pull/239/files#diff-f438134433fd9897751d1b5cd219ea33ac29be8f3af3e4ac17e65475e4cadbeeR2.

**Comment by dhh** [comment]
The chaining operator should be supported from Node 14+. I think we should probably still address this, but curious what version you're on? Maybe we need to find a way to have version requirements.

**Comment by davidrunger** [comment]
Maybe I should also mention that I am using `webpacker` (version 5.4.3), which I know has been retired. I'm not sure if that might be a factor.

**Comment by davidrunger** [comment]
Inspired by [this Stack Overflow answer](https://stackoverflow.com/a/57689106/4009384), I am able to get webpack to compile successfully by adding the following to my webpack configuration (which, for my app, is located in `config/webpack/shared.js`):

```js
// ...
module.exports = {
  // ...
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        options: {
          plugins: [
            '@babel/plugin-proposal-optional-chaining',
          ],
        },
      },
      // ...
    ],
  },
  // ...
};
```

However, doing this (i.e. adding a webpack rule for `.js` files and using babel to parse some `.js` files in `node_modules`, which I think babel doesn't do by default) seems to introduce a decently significant performance penalty (raising the `bin/webpack` time on my machine from ~40 seconds to ~56 seconds).

Additionally, it's a bit inconvenient (and I suspect not documented by `@hotwired/turbo-rails`?) that webpack users seemingly must add something like that to their webpack configuration, if it's not already there.

I'm curious to know whether I am unique in my setup and encountering this error vs. whether it is affecting / will affect many/most other users upgrading to @hotwired/turbo-rails 7.1.2. If it's a unique problem that I am having, then maybe I'll just add that to my webpack configuration and close this issude. However, if it's a problem that others are having, too, maybe the optional chaining operator added in https://github.com/hotwired/turbo-rails/pull/239 can/should be removed?

**Comment by dhh** [comment]
Yeah, easiest path here seems to be just rewriting that optional chaining use to not use that. Do take a stab if @seanpdoyle doesn't beat you to it.

**Comment by ParamagicDev** [comment]
As far as I know, this would be a Babel issue, not a Webpack issue. 

Upgrading the Babel version / Babel parser should fix this. Optionally, you can tell Webpack not to transform Turbo. This isn't even TypeScript or Node specific.

optional chaining is supported by the browser and has very good support.

https://caniuse.com/?search=optional%20chaining

just my 2 cents.

**Comment by davidrunger** [comment]
I am using the latest version of `@babel/core` (7.18.0). Here it is listed in the `yarn.lock` file of the app where I am experiencing this issue: https://github.com/davidrunger/david_runger/blob/dependabot/npm_and_yarn/hotwired/turbo-rails-7.1.2/yarn.lock#L47-L48. So I don't think that upgrading the babel version will fix this; I think that I cannot upgrade the babel version, since I am already using the latest version.

It seems that my babel version (the latest) _can_ parse the `@hotwired/turbo-rails` files, and it does when I configure webpack as mentioned in my comment above, but babel doesn't seem to do this by default (or webpack/webpacker declines to use babel to do this by default), which is seemingly for a somewhat good reason, since having babel parse all `.js` files seems to come with a transpilation performance penalty.

Telling webpack not to transform Turbo might work, but that seems to me like something that would be a little bit of friction for `@hotwired/turbo-rails` users that it'd be better if they can avoid, and also I guess that we'd lose benefits like tree shaking that come via webpack.

**Comment by ParamagicDev** [comment]
@davidrunger i believe I found the root of the problem. The parser used under the hood (acorn) didn't support optional chaining until v7.3.0.

Here's an issue thread that outlines it:

https://github.com/webpack/webpack/issues/10227

TLDR: Webpack 4 (which I believer webpacker 5.4.3 uses) doesn't support it without adding optional chaining to Babel, or setting an acorn resolution (acorn is the AST parser used under the hood).
